DeepNtuplizer/plugins/DeepNtuplizer.cc

```diff
@@ -191,59 +199,78 @@ DeepNtuplizer::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup)
   if (vertices->empty()) return; // skip the event if no PV found
   const reco::Vertex &PV = vertices->front();
 
-  edm::Handle<pat::JetCollection> jets;
+  edm::Handle<edm::View<pat::Jet> > jets;
   iEvent.getByToken(jetToken_, jets);
 
+  edm::Handle<edm::Association<reco::GenJetCollection> > genJetMatchRecluster;
+  iEvent.getByToken(genJetMatchReclusterToken_, genJetMatchRecluster);
+
+  edm::Handle<edm::Association<reco::GenJetCollection> > genJetMatchWithNu;
+  iEvent.getByToken(genJetMatchWithNuToken_, genJetMatchWithNu);
+
   // clear vectors
   npv_ = vertices->size();
 
 
   // loop over the jets
-  for (const pat::Jet &jet : *jets) {
+  for (edm::View<pat::Jet>::const_iterator itJet = jets->begin(); itJet != jets->end(); itJet++) {
+    unsigned int idx = itJet - jets->begin();       
 
    // some cuts to contrin training region 
-    if ( jet.pt() < jetPtMin_ ||  jet.pt() > jetPtMax_ ) continue;                  // apply jet pT cut
-    if ( jet.eta() < fabs(jetAbsEtaMin_) ||jet.eta() > fabs(jetAbsEtaMax_) ) continue; // apply jet eta cut
+    if ( itJet->pt() < jetPtMin_ ||  itJet->pt() > jetPtMax_ ) continue;                  // apply jet pT cut
+    if ( itJet->eta() < fabs(jetAbsEtaMin_) ||itJet->eta() > fabs(jetAbsEtaMax_) ) continue; // apply jet eta cut
     // often we have way to many gluons that we do not need. This randomply reduces the gluons
     if (gluonReduction_>0)
       {
-	if(jet.partonFlavour()==21)
+	if(itJet->partonFlavour()==21)
 	  {
 	    if(TRandom_.Uniform()>gluonReduction_) continue; 
 	  }
       }
 
     // truth labels
     gen_pt_ = 0.;
-    if(jet.genJet()!=NULL)   gen_pt_ =  jet.genJet()->pt();
-    isB_= int(abs(jet.partonFlavour())==5);
-    isC_= int(abs(jet.partonFlavour())==4);
-    isUDS_= int( (abs(jet.partonFlavour())>0) && (abs(jet.partonFlavour())<4 ));
-    isG_= int(jet.partonFlavour()==21);
-
-    jet_pt_ = jet.pt();
-    jet_eta_ = jet.eta();
+    if(itJet->genJet()!=NULL)   gen_pt_ =  itJet->genJet()->pt();
+    const edm::RefToBase<pat::Jet> patJetRef = jets->refAt(idx);      
```

Comment:
Changed token from pat::JetCollection to edm::View<pat::Jet> in order to be able to get refAt. A bit awkward to change from const pat::Jet &jet to iterator within the loop. Suggestions to keep the old loop (and still get the reference to access the matching) are welcome 👍